### PR TITLE
support find stores with tiflash_mpp label

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

close： #498

Related tidb pr: https://github.com/pingcap/tidb/pull/33535

1. add `GetTiFlashMPPRPCContextByConsistentHash` and its related functions.
2. handle tiflash_mpp in `onSendFail()`.
3. change some if branch from `ctx.Store.storeType == tikvrpc.TiFlash` `to ctx.Store.storeType.IsTiFlashRelatedType()`.

### CODE REVIEW

File: internal/locate/region_cache.go [hunk 1]
```diff
@@ -1723,6 +1730,59 @@ func (c *RegionCache) GetTiFlashStores() []*Store {
...
+	for _, s := range stores {
+		tmpStore := &Store{
+			storeID:   s.GetId(),
+			addr:      s.GetAddress(),
+			saddr:     s.GetStatusAddress(),
+			storeType: tikvrpc.GetStoreTypeByMeta(s),
+			labels:    s.GetLabels(),
+			state: uint64(resolved),
+		}
+		if tmpStore.IsLabelsMatch(mppLabels) {
```
Comment (XuHuaiyu) [inline internal/locate/region_cache.go:1766]: Will this be better?
``` go
if s.IsLableMatch(mppLabels) {
    tmpStore = &Store{}
    res = append(res, tmpStore)
}
```

File: internal/locate/region_cache.go [hunk 2]
```diff
@@ -1723,6 +1730,59 @@ func (c *RegionCache) GetTiFlashStores() []*Store {
...
+
+func (c *RegionCache) reloadTiFlashMPPStores(bo *retry.Backoffer) (res []*Store, _ error) {
+	stores, err := c.pdClient.GetAllStores(bo.GetCtx())
+	if err != nil {
+		return nil, err
+	}
+	mppLabels := []*metapb.StoreLabel{
+		{
+			Key:   "engine",
+			Value: "tiflash_mpp",
```
Comment (you06) [inline internal/locate/region_cache.go:1754]: Better to use constants, I find some [similar usage](https://github.com/tikv/client-go/blob/79f010d806110cb4bccf9218ceae5e5db54951fe/tikvrpc/endpoint.go#L62-L67) here.

Comment (you06) [inline internal/locate/region_cache.go:1754]: Better to use some constants, I find some [similar usage](https://github.com/tikv/client-go/blob/79f010d806110cb4bccf9218ceae5e5db54951fe/tikvrpc/endpoint.go#L62-L67) here.

File: internal/locate/region_cache.go [hunk 3]
```diff
@@ -389,6 +394,8 @@ func NewRegionCache(pdClient pd.Client) *RegionCache {
 	c.mu.latestVersions = make(map[uint64]RegionVerID)
 	c.mu.sorted = btree.New(btreeDegree)
 	c.storeMu.stores = make(map[uint64]*Store)
+	c.tiflashMPPStoreMu.needReload = true
+	c.tiflashMPPStoreMu.stores = make([]*Store, 1)
```
Comment (you06) [inline internal/locate/region_cache.go:398]: It'll be reloaded when reading for the first time, why initiate it with a non-zero-length array?

Comment (you06) [inline internal/locate/region_cache.go:398]: It'll be reloaded when reading for the first time, why initiate it with a non-zero-length array?

File: internal/locate/region_cache.go [hunk 4]
```diff
@@ -1723,6 +1730,59 @@ func (c *RegionCache) GetTiFlashStores() []*Store {
...
 }
 
+// GetTiFlashMPPStores returns all stores with lable <engine, tiflash_mpp>.
+func (c *RegionCache) GetTiFlashMPPStores(bo *retry.Backoffer) (res []*Store, err error) {
+	c.tiflashMPPStoreMu.RLock()
+	needReload := c.tiflashMPPStoreMu.needReload
+	stores := c.tiflashMPPStoreMu.stores
+	c.tiflashMPPStoreMu.RUnlock()
+
+	if needReload {
```
Comment (you06) [inline internal/locate/region_cache.go:1740]: So in a bad case, there may be many threads calling this function and turned to reload, however one reload task is enough.

Comment (you06) [inline internal/locate/region_cache.go:1740]: So in a bad case, there may be many threads calling this function and turned to reload, however one reload task is enough.

File: internal/locate/region_cache.go (hunk 1 above)
Comment (guo-shaoge) [reply internal/locate/region_cache.go]: Add `isStoreContainLabel`.

File: internal/locate/region_cache.go (hunk 2 above)
Comment (guo-shaoge) [reply internal/locate/region_cache.go]: Done

File: internal/locate/region_cache.go (hunk 3 above)
Comment (guo-shaoge) [reply internal/locate/region_cache.go]: done

File: internal/locate/region_cache.go (hunk 4 above)
Comment (guo-shaoge) [reply internal/locate/region_cache.go:1792]: Yes. But looks like `RegionCache` is the same behavior?

Comment (you06) [reply internal/locate/region_cache.go:1792]: Ok, this looks good to me.

**Comment by you06** [comment]
What about adding [counter and duration metrics](https://github.com/tikv/client-go/blob/5d3195ce1174d3fd918c00281e77fdae078db81c/metrics/shortcuts.go#L193-L208) for the cache invalidation and loading.

**Comment by guo-shaoge** [comment]
> What about adding [counter and duration metrics](https://github.com/tikv/client-go/blob/5d3195ce1174d3fd918c00281e77fdae078db81c/metrics/shortcuts.go#L193-L208) for the cache invalidation and loading.

To keep this pr clean, I will update metrics in the next pr(https://github.com/tikv/client-go/issues/509)~

**Review by you06 — APPROVED** [review]

**Review by XuHuaiyu — APPROVED** [review]

**Review by sticnarf — APPROVED** [review]

**Pull request merged**